### ISSUE TITLE
Special marks change on diffupdate

### ISSUE THREAD

**Issue by qstrahl** [neovim/neovim#4653]
- Neovim version: 0.1.4-dev
- [ ] Vim behaves differently? Vim version:
- Operating system/version: Ubuntu 15.10
- Terminal name/version: rxvt-unicode 9.21
- `$TERM`: tmux-256color

### Actual behaviour
When editing buffers in "diff mode", a `:diffupdate` causes the ``` `[ ``` and ``` `] ``` marks to be moved to the first and last lines of the buffer. (Vim apparently has this bug, too)

### Expected behaviour
`:diffupdate` should not change these marks.

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
:e file1
:vsp file2
:windo diffthis
yy
`[
`]
:diffupdate
`[
`]
```

**Comment by qstrahl** [comment]
So I've narrowed this down to the [diff_write](/neovim/neovim/blob/d2cbc311855114b7736919a344a82e303d4a8164/src/nvim/diff.c#L627-L636) function:
```c
static int diff_write(buf_T *buf, char_u *fname)
{
  char_u *save_ff = buf->b_p_ff;
  buf->b_p_ff = vim_strsave((char_u *)FF_UNIX);
  int r = buf_write(buf, fname, NULL, (linenr_T)1, buf->b_ml.ml_line_count,
                    NULL, FALSE, FALSE, FALSE, TRUE);
  free_string_option(buf->b_p_ff);
  buf->b_p_ff = save_ff;
  return r;
}
```

Note this line in particular:
```c
int r = buf_write(buf, fname, NULL, (linenr_T)1, buf->b_ml.ml_line_count,
                    NULL, FALSE, FALSE, FALSE, TRUE);
```

This is the culprit. Looking at [this part of buf_write](/neovim/neovim/blob/de1084f3c48816e96be3cac28bcf56bd321ad800/src/nvim/fileio.c#L2360-L2363):
```c
  /*
   * Set '[ and '] marks to the lines to be written.
   */
  buf->b_op_start.lnum = start;
  buf->b_op_start.col = 0;
  buf->b_op_end.lnum = end;
  buf->b_op_end.col = 0;
```

There doesn't seem to be a way to avoid this block of code, and I'm not familiar enough with the code base to say how best to fix this. Another argument to `buf_write` which allows us to specify or preserve the `'[` and `']` marks, maybe?

**Comment by qstrahl** [comment]
It looks like there's at least one other place where this issue has been identified and circumvented; [do_filter in ex_cmds.c](/neovim/neovim/blob/d2cbc311855114b7736919a344a82e303d4a8164/src/nvim/ex_cmds.c#L1227-L1236):
```c
      /*
       * Put cursor on first filtered line for ":range!cmd".
       * Adjust '[ and '] (set by buf_write()).
       */
      curwin->w_cursor.lnum = line1;
      del_lines(linecount, TRUE);
      curbuf->b_op_start.lnum -= linecount;             /* adjust '[ */
      curbuf->b_op_end.lnum -= linecount;               /* adjust '] */
      write_lnum_adjust(-linecount);                    /* adjust last line
                                                           for next write */
```

So we could copy this but given that there are now multiple places where this is needed, it would probably be better to encapsulate the behaviour in one place. I _feel_ like the cleanest way is to add a flag to `buf_write` but I would appreciate the input of someone who actually works with the code. :D

**Comment by jamessan** [comment]
> It looks like there's at least one other place where this issue has been addressed; do_filter in ex_cmds.c:

That's handling a different issue.  When you filter a part of the buffer through an external command, the number of lines may change.  This code is handling that adjustment.

**Comment by qstrahl** [comment]
Ah, I see. Thanks for the clarification. Do you have any opinion on how we should handle it in `diff_write`?

**Comment by jamessan** [comment]
As a workaround, does `:lockmarks diffupdate` help?

**Comment by qstrahl** [comment]
It does not, unfortunately. `lockmarks` does not lock `'[` and `']` :<

**Comment by jamessan** [comment]
Yeah, I had a recollection it didn't, yet one of the plugins distributed in Vim uses `lockmarks` explicitly to protect those marks...

I'll poke around and see what comes to mind.

**Comment by jamessan** [comment]
I have a possible solution but this whole thing is a mess since there isn't distinct storage for `the '[ and '] marks` vs. `the range of an operation`.  That's really convenient in one sense, because you just use the latter to represent the former.  When you use `yy`, the range of that operation is the current line, so the marks are set to that line.

The problem is that since those two concepts are conflated, you have to play whack-a-mole with figuring out what should be affected by `:lockmarks`.

My current patch _only_ adds knowledge of `:lockmarks` to `buf_write()` and `readfile()`.  Then fixing the diff commands is easy -- temporarily enable `lockmarks` in `diff_write()`.

I'll submit a PR for this limited change upstream, once I write some tests, and see what Bram says.
